include/ringmesh/basic/attributes.h

```diff
@@ -0,0 +1,1024 @@
+        *  can be overloaded in derived classes.
+        */
+        void apply_permutation(
+            const std::vector< index_t >& permutation ) final
+        {
+            std::vector< T > tmp_vector;
+            tmp_vector.resize( vector_.size() );
+            for( auto p : range( permutation.size() ) )
+            {
+                tmp_vector[p] = vector_[permutation[p]];
+            }
+            vector_ = tmp_vector;
+        }
+
+    protected:
+        std::vector< T > vector_;
+    };
+
+    template < class T >
+    class VectorStore : public VectorStoreBase< T >
+    {
+    public:
+        VectorStore() = default;
+
+        Store* clone() const final
+        {
+            VectorStore< T >* result = new VectorStore< T >();
+            result->resize( this->size() );
+            result->vector_ = this->vector_;
+            return result;
+        }
+    };
+
+    template <>
+    class VectorStore< bool > : public VectorStoreBase< Byte >
+    {
+    public:
+        /**
+        * \brief Creates a new empty attribute store.
+        */
+        VectorStore() = default;
+
+        Store* clone() const final
+        {
+            VectorStore< bool >* result = new VectorStore< bool >;
+            result->resize( this->size() );
+            result->vector_ = this->vector_;
+            return result;
+        }
+        std::string element_typeid_name() const final
+        {
+            return typeid( bool ).name();
+        }
+    };
+
+    template < class T >
+    class ConstantStore : public TypedStore< T >
+    {
+    public:
+        /**
+        * \brief Creates a new empty attribute store.
+        */
+        ConstantStore()
+        {
+            this->size_ = 1;
+        }
+
+        void resize( index_t ) final
+        {
+            Logger::warn( "Trying to resize constant attribute" );
+        }
+
+        const void* data() const final
+        {
+            return &constant_;
+        }
+        void* data() final
+        {
+            return &constant_;
+        }
+        const void* element( index_t element ) const final
+        {
+            ringmesh_unused( element );
+            return &this->constant_;
+        }
+        void* element( index_t element ) final
+        {
+            ringmesh_unused( element );
+            return &this->constant_;
+        }
+        void clear() final
+        {
+        }
+
+        Store* clone() const final
+        {
+            ConstantStore< T >* result = new ConstantStore< T >();
+            result->constant_ = constant_;
+            return result;
+        }
+        void compress( const std::vector< index_t >& old2new ) final
+        {
+            ringmesh_unused( old2new );
+        }
+
+        void apply_permutation(
+            const std::vector< index_t >& permutation ) final
+        {
+            ringmesh_unused( permutation );
+        }
+
+    protected:
+        T constant_;

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)